### PR TITLE
Support setting max contacts in dart's bullet collision detector 

### PR DESCRIPTION
**Pull request by iche033**
# 🎉 New feature

depends on #584

## Summary

This PR extends  dart's bullet collision detector to support the max contact collision pair feature.  

This is a follow-up to #582 which added this feature for dart's ode collision detector. 

Changes:
* Refactored existing implementation to derive off of a base `GzCollisionDetector` a class for limiting max contact points in dartsim: 4da0d6f20c699d1e86b2380afad0064d4f89a919
* Added a new `GzBulletCollisionDetector` class for extending the max contact feature to dart's bullet collision detector: 4da0d6f20c699d1e86b2380afad0064d4f89a919

## Test it

Updated test in  `simulation_features.c` to test the max contact feature with dart's bullet collision detector

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: dartsim/src/GzCollisionDetector.hh [hunk 1]
```diff
@@ -15,15 +15,50 @@
...
+  /// \brief Set the maximum number of contacts between a pair of collision
+  /// objects
+  /// \param[in] _maxContacts Maximum number of contacts between a pair of
+  /// collision objects.
+  public: virtual void SetCollisionPairMaxContacts(std::size_t _maxContacts);
+
+  /// \brief Get the maximum number of contacts between a pair of collision
+  /// objects
+  /// \return Maximum number of contacts between a pair of collision objects.
+  public: virtual std::size_t GetCollisionPairMaxContacts() const;
```
Comment (ahcorde) [inline dartsim/src/GzCollisionDetector.hh:42]: include `<cstdio>`

**Review by ahcorde — CHANGES_REQUESTED** [review]

File: dartsim/src/GzCollisionDetector.hh (hunk 1 above)
Comment (iche033) [reply dartsim/src/GzCollisionDetector.hh:43]: added in [fa9468f](https://github.com/gazebosim/gz-physics/pull/593/commits/fa9468f8b95a886bcc856aaf3aa0f94625625e65)

**Comment by azeey** [comment]
@azeey to review

**Comment by iche033** [comment]
> Looks great to me! Sorry for the long delay in reviewing this. One unfortunate consequence is that Garden is now EOL, so we shouldn't merge into this branch. Do you mind retargetting this to gz-physics8(Ionic)?

retargeted to gz-physics8

File: dartsim/src/WorldFeatures.cc [hunk 2]
```diff
@@ -114,8 +114,18 @@ void WorldFeatures::SetWorldCollisionPairMaxContacts(
...
+      collisionDetector);
+    if (bulletCollisionDetector)
+    {
+      bulletCollisionDetector->SetCollisionPairMaxContacts(_maxContacts);
+    }
+    else
+    {
+      gzwarn << "Currently max contacts feature is only supported by the "
+             << "bullet and ode collision detector in dartsim." << std::endl;
+    }
```
[the remark this answers is not in the document]
Comment (azeey) [reply dartsim/src/WorldFeatures.cc]: I think the same can be used below.

[the remark this answers is not in the document]
Comment (iche033) [reply dartsim/src/WorldFeatures.cc]: oops yes, missed that one, updated [f708237](https://github.com/gazebosim/gz-physics/pull/593/commits/f708237a61aa5202e72d6e45405db2fe90942ecd)

**Review by azeey — APPROVED** [review]

**Comment by iche033** [comment]
@mergifyio backport gz-physics7

**Pull request merged**